File: common/src/main/java/de/hype/bingonet/shared/objects/SplashData.java
package de.hype.bingonet.shared.objects;

import de.hype.bingonet.shared.constants.Islands;
import de.hype.bingonet.shared.constants.StatusConstants;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class SplashData {
    public String announcer;
    public StatusConstants status;
    public int splashId;
    public SplashLocation locationInHub;

    public String extraMessage;
    public boolean lessWaste;
    public String serverID;
    /**
     * If null the Splash is in a private Mega → Request Invite.
     */
    @Nullable
    public HubSelectorData hubSelectorData;

    public SplashData(String user, SplashLocation locationInHub, String extraMessage, boolean lessWaste, String serverID, @Nullable HubSelectorData hubSelectorData, StatusConstants status) {
        this.serverID = serverID;
        this.announcer = user;
        this.locationInHub = locationInHub;
        this.extraMessage = extraMessage != null ? extraMessage.replace("&", "§") : null;
        this.lessWaste = lessWaste;
        this.status = status;
        this.hubSelectorData = hubSelectorData;
    }

    public SplashData(String user, SplashLocation locationInHub, String extraMessage, boolean lessWaste, String serverID, @Nullable HubSelectorData hubSelectorData) {
        this(user, locationInHub, extraMessage, lessWaste, serverID, hubSelectorData, StatusConstants.WAITING);
    }

    public SplashData(SplashData packet) {
        this(packet.announcer, packet.locationInHub, packet.extraMessage, packet.lessWaste, packet.serverID, packet.hubSelectorData, packet.status);
        this.splashId = packet.splashId;
    }


    public static class HubSelectorData {
        public int hubNumber;
        public Islands hubType;

        public HubSelectorData(int hubNumber, @NotNull Islands hubType) {
            if (!(hubType.equals(Islands.HUB) || hubType.equals(Islands.DUNGEON_HUB))) {
                throw new IllegalArgumentException("§cInvalid hub type specified. Please only use the Suggestions!");
            }
            if (hubNumber < 1 || hubNumber > 28) {
                throw new IllegalArgumentException("§cInvalid hub number specified. Must be between 1 and 28");
            }
            this.hubNumber = hubNumber;
            this.hubType = hubType;
        }
    }
}

File: common/src/main/java/de/hype/bingonet/shared/objects/SplashData.kt
package de.hype.bingonet.shared.objects

import de.hype.bingonet.shared.constants.Islands
import de.hype.bingonet.shared.constants.StatusConstants

open class SplashData @JvmOverloads constructor(
    @JvmField var announcer: String,
    @JvmField var locationInHub: SplashLocation,
    extraMessage: String?,
    @JvmField var lessWaste: Boolean,
    @JvmField var serverID: String?,
    /**
     * If null the Splash is in a private Mega → Request Invite.
     */
    @JvmField var hubSelectorData: HubSelectorData?,
    @JvmField var status: StatusConstants = StatusConstants.WAITING
) {
    @JvmField
    var splashId: Int = 0

    @JvmField
    var extraMessage: String?

    init {
        this.extraMessage = extraMessage?.replace("&", "§")
    }

    constructor(packet: SplashData) : this(
        packet.announcer,
        packet.locationInHub,
        packet.extraMessage,
        packet.lessWaste,
        packet.serverID,
        packet.hubSelectorData,
        packet.status
    ) {
        this.splashId = packet.splashId
    }

    fun update(data: SplashData) {
        this.announcer = data.announcer
        this.locationInHub = data.locationInHub
        this.extraMessage = data.extraMessage
        this.lessWaste = data.lessWaste
        this.serverID = data.serverID
        this.hubSelectorData = data.hubSelectorData
        this.status = data.status
        this.splashId = data.splashId
    }


    class HubSelectorData(hubNumber: Int, hubType: Islands) {
        @JvmField
        var hubNumber: Int

        @JvmField
        var hubType: Islands?

        init {
            require(hubType == Islands.HUB || hubType == Islands.DUNGEON_HUB) { "§cInvalid hub type specified. Please only use the Suggestions!" }
            require(!(hubNumber < 1 || hubNumber > 28)) { "§cInvalid hub number specified. Must be between 1 and 28" }
            this.hubNumber = hubNumber
            this.hubType = hubType
        }
    }
}